src/coreclr/vm/debugdebugger.cpp

```diff
@@ -230,43 +226,221 @@ FCIMPL0(FC_BOOL_RET, DebugDebugger::IsLogging)
 }
 FCIMPLEND
 
-FCIMPL4(void, DebugStackTrace::GetStackFramesInternal,
-        StackFrameHelper* pStackFrameHelperUNSAFE,
-        INT32 iSkip,
-        FC_BOOL_ARG fNeedFileInfo,
-        Object* pExceptionUNSAFE
-       )
+static StackWalkAction GetStackFramesCallback(CrawlFrame* pCf, VOID* data)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    DebugStackTrace::GetStackFramesData* pData = (DebugStackTrace::GetStackFramesData*)data;
+
+    if (pData->skip > 0)
+    {
+        pData->skip--;
+        return SWA_CONTINUE;
+    }
+
+    // <REVISIT_TODO>@todo: How do we know what kind of frame we have?</REVISIT_TODO>
+    //        Can we always assume FramedMethodFrame?
+    //        NOT AT ALL!!!, but we can assume it's a function
+    //                       because we asked the stackwalker for it!
+    MethodDesc* pFunc = pCf->GetFunction();
+
+    if (pData->cElements >= pData->cElementsAllocated)
+    {
+
+        DebugStackTrace::Element* pTemp = new (nothrow) DebugStackTrace::Element[2*pData->cElementsAllocated];
+
+        if (!pTemp)
+        {
+            return SWA_ABORT;
+        }
+
+        memcpy(pTemp, pData->pElements, pData->cElementsAllocated * sizeof(DebugStackTrace::Element));
+
+        delete [] pData->pElements;
+
+        pData->pElements = pTemp;
+        pData->cElementsAllocated *= 2;
+    }
+
+    PCODE ip;
+    DWORD dwNativeOffset;
+
+    if (pCf->IsFrameless())
+    {
+        // Real method with jitted code.
+        dwNativeOffset = pCf->GetRelOffset();
+        ip = GetControlPC(pCf->GetRegisterSet());
+    }
+    else
+    {
+        ip = (PCODE)NULL;
+        dwNativeOffset = 0;
+    }
+
+    // Pass on to InitPass2 that the IP has already been adjusted (decremented by 1)
+    INT flags = pCf->IsIPadjusted() ? STEF_IP_ADJUSTED : 0;
+
+    pData->pElements[pData->cElements].InitPass1(
+            dwNativeOffset,
+            pFunc,
+            ip,
+            flags);
+
+    // We'll init the IL offsets outside the TSL lock.
+
+    ++pData->cElements;
+
+    // Since we may be asynchronously walking another thread's stack,
+    // check (frequently) for stack-buffer-overrun corruptions after
+    // any long operation
+    pCf->CheckGSCookies();
+
+    // check if we already have the number of frames that the user had asked for
+    if ((pData->NumFramesRequested != 0) && (pData->NumFramesRequested <= pData->cElements))
+    {
+        return SWA_ABORT;
+    }
+
+    return SWA_CONTINUE;
+}
+
+static void GetStackFrames(
```

Comment:
Converted from member function to private in translation unit. This was moved from below.